### PR TITLE
Compatibility with docker secrets

### PR DESCRIPTION
**Pull request by ekho**
Read passwords from files specified with *_PASSWORD_FILE variables

### CODE REVIEW

File: docker-entrypoint.sh [hunk 1]
```diff
@@ -19,6 +19,32 @@ run_as() {
...
+	if [ "${!var:-}" ] && [ "${!fileVar:-}" ]; then
+		echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
+		exit 1
+	fi
+	local val=""
+	if [ "${!var:-}" ]; then
+		val="${!var}"
+	elif [ "${!fileVar:-}" ]; then
+		val="$(< "${!fileVar}")"
+    	unset "$fileVar"
```
Comment (J0WI) [inline docker-entrypoint.sh]: code style

Comment (ekho) [reply docker-entrypoint.sh]: Where can I read about recommended code style?

Comment (J0WI) [reply docker-entrypoint.sh]: Just use the same code style in the whole file.
The `unset` statement is either not intended correctly or it should be placed after the conditions, near `export "$var"="$val"`.

Comment (ekho) [reply docker-entrypoint.sh]: `unset` used to remove `*_PASSWORD_FILE` variable when it defined. Why should I unset this variable while it is not defined?

Comment (ekho) [reply docker-entrypoint.sh]: This block of code got from [official mariadb Dockerfile](https://github.com/docker-library/mariadb/blob/master/10.3/docker-entrypoint.sh#L21-L41)

Comment (J0WI) [reply docker-entrypoint.sh]: The line indention there looks correctly.

**Comment by J0WI** [comment]
LGTM
Would be great if you could also open a PR to update the content here: https://github.com/docker-library/docs/blob/master/nextcloud/content.md

**Comment by dargmuesli** [comment]
There are a few merge commits that need to be resolved. I'd really like to see this feature implemented! :)

**Comment by J0WI** [comment]
Merged in #729, thanks!

**Comment by epma01** [comment]
invalid path, causing not working "New nextcloud instance"

```
+ echo 'New nextcloud instance'
New nextcloud instance
+ file_env NEXTCLOUD_ADMIN_PASSWORD
+ local 'var=NEXTCLOUD_ADMIN_PASSWORD'
+ local 'fileVar=NEXTCLOUD_ADMIN_PASSWORD_FILE'
+ local 'def='
./entrypoint.sh: line 171: syntax error: bad substitution
```

Introduced code is using bash indirect expansion but entrypoint.sh is not using bash and it is not supported
```
file_env() {
    local var="$1"
    local fileVar="${var}_FILE"
    local def="${2:-}"
>>>    if [ "${!var:-}" ] && [ "${!fileVar:-}" ]; then    <<< this line is causing the issue
        echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
        exit 1
    fi
```

**Comment by SnowMB** [comment]
Reopen because of revert in #733

**Comment by ekho** [comment]
@J0WI I fixed #731

File: README.md [hunk 2]
```diff
@@ -245,6 +245,65 @@ services:
...
+      - POSTGRES_PASSWORD_FILE=/run/secrets/postgres_password
+      - NEXTCLOUD_ADMIN_USER=superuser
+      - NEXTCLOUD_ADMIN_PASSWORD_FILE=/run/secrets/admin_password
+    depends_on:
+      - db
+    secrets:
+      - postgres_password
+      - admin_password
+
+  cron:
```
Comment (J0WI) [inline README.md]: I'd prefer to keep this example simple and without cron.

**Comment by J0WI** [comment]
I'm still having no success with https://github.com/nextcloud/docker/pull/560/commits/2177e240c669e58b204b874f984208ef030028dd.

**Comment by ekho** [comment]
@J0WI I have no docker swarm now. But I tested it this way:
```
echo 'PASSw0rd' > /tmp/nx-adm-pass
cd 16.0/apache
docker build . -t nextcloud:apache-secrets-test
docker run --rm -it -e 'NEXTCLOUD_ADMIN_PASSWORD_FILE=/run/secrets/admin-password' -e 'NEXTCLOUD_UPDATE=1' -v /tmp/nx-adm-pass:/run/secrets/admin-password nextcloud:apache-secrets-test sh -c 'echo "admin password is $NEXTCLOUD_ADMIN_PASSWORD"'
```
and output:
```
Initializing nextcloud 16.0.2.1 ...
Initializing finished
New nextcloud instance
admin password is PASSw0rd
```

**Comment by J0WI** [comment]
That worked for me. But therefore we should not include a `docker-compose` example.

**Comment by dargmuesli** [comment]
Why was this closed?

**Comment by ekho** [comment]
@dargmuesli I don't have any time to work on this pr

**Comment by dargmuesli** [comment]
Is there anything missing I could help with?

**Comment by ekho** [comment]
@dargmuesli I can grant contributor role for you on my repo and you can fix issues

**Comment by dargmuesli** [comment]
Yep, that'd be alright. I didn't follow this in detail, but if someone needs me to fix issues, I'd be happy to help.

**Comment by ekho** [comment]
@dargmuesli Thank you. I added you to my fork

**Comment by dargmuesli** [comment]
Thank you :)

**Comment by dargmuesli** [comment]
Squashed & rebased. Works with my test swarm.
@J0WI could you verify, please? :)

**Comment by dargmuesli** [comment]
Because the Postgres image's description [states about its Docker Secret support](https://github.com/docker-library/docs/tree/master/postgres#docker-secrets):

> Currently, this is only supported for POSTGRES_INITDB_ARGS, POSTGRES_PASSWORD, POSTGRES_USER, and POSTGRES_DB

The nextcloud administrator user name should be secretly settable too (imo), to add further security. Having to guess the admin's user name makes brute force attacks much more difficult.

I'll add the appropriate `MYSQL_` variables too and update the README.

**Comment by dargmuesli** [comment]
Done.

**Comment by J0WI** [comment]
@tilosp can I ask you for a second review?

**Comment by andrasmaroy** [comment]
While adding secrets shouldn't SMTP_PASSWORD also added this way?

**Comment by dargmuesli** [comment]
`SMTP_PASSWORD` is not read in the entrypoint script, but in a php script instead. I think another PR would be appropriate for this variable's secret support. I want this PR to be merged first though, as it's open for quite some time now ;)

**Pull request merged**